ext/ddtrace_profiling_native_extension/heap_recorder.c

```diff
@@ -172,7 +201,13 @@ void heap_recorder_iterate_stacks(heap_recorder *heap_recorder, void (*for_each_
   internal_iteration_data.for_each_callback = for_each_callback;
   internal_iteration_data.for_each_callback_extra_arg = for_each_callback_extra_arg;
   internal_iteration_data.heap_recorder = heap_recorder;
-  st_foreach(heap_recorder->heap_records, st_heap_records_iterate, (st_data_t) &internal_iteration_data);
+  if (heap_recorder->enable_heap_size_profiling) {
+    // To get an accurate object size, we need to query our tracked live objects so lets iterate over the object_records
+    st_foreach(heap_recorder->object_records, st_object_records_iterate, (st_data_t) &internal_iteration_data);
+  } else {
+    // If we're just reporting heap counts/samples, it's faster to iterate on the heap records
+    st_foreach(heap_recorder->heap_records, st_heap_records_iterate, (st_data_t) &internal_iteration_data);
+  }
```

Comment:
Uh, this is interesting -- may be worth calling out as an intro comment in the design of the heap recorder at some point.